modules/images/webp-uploads/load.php

```diff
@@ -35,5 +35,170 @@ function webp_uploads_filter_image_editor_output_format( $output_format, $filena
 
 	return $output_format;
 }
-add_filter( 'image_editor_output_format', 'webp_uploads_filter_image_editor_output_format', 10, 3 );
 
+/**
+ * Hook called by `wp_generate_attachment_metadata` to create the `sources` property for every image
+ * size, the sources' property would create a new image size with all the mime types specified in
+ * `webp_uploads_valid_image_mime_types`. If the original image is one of the mimes from
+ * `webp_uploads_valid_image_mime_types` the image is just added to the `sources` property and  not
+ * created again. If the uploaded attachment is not a valid image this function does not alter the
+ * metadata of the attachment, on the other hand a `sources` property is added.
+ *
+ * @since n.e.x.t
+ *
+ * @see   wp_generate_attachment_metadata
+ * @see   webp_uploads_valid_image_mime_types
+ *
+ * @param array $metadata      An array with the metadata from this attachment.
+ * @param int   $attachment_id The ID of the attachment where the hook was dispatched.
+ *
+ * @return array An array with the updated structure for the metadata before is stored in the database.
+ */
+function webp_uploads_create_sources_property( array $metadata, $attachment_id ) {
+	// This should take place only on the JPEG image.
+	$valid_mime_types = webp_uploads_valid_image_mime_types();
+
+	// Not a supported mime type to create the sources property.
+	if ( ! array_key_exists( get_post_mime_type( $attachment_id ), $valid_mime_types ) ) {
+		return $metadata;
+	}
+
+	// All subsizes are created out of the `file` property.
+	$file = get_attached_file( $attachment_id, true );
+
+	// File does not exist.
+	if ( ! file_exists( $file ) ) {
+		return $metadata;
+	}
+
+	// Prevent to convert JPEG to WebP if we are creating JPEG versions of the image.
+	remove_filter( 'image_editor_output_format', 'webp_uploads_filter_image_editor_output_format' );
+	$backup_sizes = get_post_meta( $attachment_id, '_wp_attachment_backup_sizes', true );
+	$backup_sizes = is_array( $backup_sizes ) ? $backup_sizes : array();
+
+	foreach ( webp_uploads_get_image_sizes() as $size => $properties ) {
```

Comment:
instead of trying to recreate sizes, can we just copy the meta 'sizes'?

Reply:
We can certainly do that but aside from that, we need to know if the newly created image should be `cropped` or not and that's the main reason why we need this function. 

I think something we can do at the top of the `foreach` loop to make sure this `$size` exists in the original metadata before creating a targeted image instead.

Reply:
Actually, this is already the case on this line :smiley:  

- https://github.com/WordPress/performance/pull/154/files/84eb3ec30e532939cf81c65d4fc46120ae1aac8c#diff-71ed188a16b79d179868ee80e76f775462a451e871fc3d6214fcacca3edcea8cR81

Reply:
One thing I noticed in testing when I uploading a very large image, I got a jpeg created with a `-scaled` suffix, then **all** of the generated `.webp` sub size files also included the `-scaled` suffix:

![image](https://user-images.githubusercontent.com/2676022/154143125-397cc077-76ca-40e3-a57f-abef530a206c.png)

In addition, I noticed there was no full sized `.webp` created (that is `{filename}-scaled.webp`) - we need that full sized file for the image markup (only applies when you upload images larger than `big_image_size_threshold`)

Reply:
Thanks for testing and for the feedback. Some follow-up comments.

> One thing I noticed in testing when I uploading a very large image, I got a jpeg created with a -scaled suffix, then all of the generated .webp sub size files also included the -scaled suffix.

Great point mostly because this is created out of the scaled version. But let me fix this one :) 

> In addition, I noticed there was no full sized .webp created (that is {filename}-scaled.webp) - we need that full sized file for the image markup (only applies when you upload images larger than big_image_size_threshold)

Actually, this follows the same logic applied here #122  and #143   this behavior would go against this solved issue, as this `scaled` version is the `full` size image. Based on the issue `scaled` and `full` size always remains in the original image format. (not converted)

Let me know if you think we should change it instead.